### ISSUE TITLE
Caching text features in memory may increase the training speed of YOLO World models by four times.

### ISSUE THREAD

**Issue by h13-0** [ultralytics/ultralytics#20469]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

# Description
In YOLO World training, each batch uses clip to calculate text features for **each training sample** before training, and then normalizes and trains the text features.

In the comments of `WorldTrainerFromScratch`, the training set is selected as Object 365 and flickr30k. In the Objectis365 dataset, there are over 30 million samples, but only 365 different labels; The flickr30k dataset has approximately 150000 samples, but only a few thousand different text prompts. This also means that tens of millions of repeated text features need to be calculated at each epoch stage.

In this case, although the YOLO World model itself does not add too many parameters, its training speed is significantly slower than ordinary models.

# Test condition
- GPU: RTX3090
- Model: YOLOV8S-Worldv2
- Dataset:
    - Train sets:
        - coco
        - flickr30k
    - Valid sets:
        - lvis

## Test code
```python
from ultralytics.models.yolo.world.train_world import WorldTrainerFromScratch
from ultralytics import YOLOWorld

def main():
    data = dict(
        train=dict(
            yolo_data=["coco.yaml"],
            grounding_data=[
                dict(
                    img_path="../datasets/flickr30k/flickr30k-images",
                    json_file="../datasets/flickr30k/flickr30k-json/final_flickr_separateGT_train.json",
                ),
            ],
        ),
        val=dict(yolo_data=["lvis.yaml"]),
    )

    model = YOLOWorld("yolov8s-worldv2.yaml")
    model.train(data=data, trainer=WorldTrainerFromScratch, epochs=300, imgsz=640, batch=36)

if __name__ == '__main__':
    main()
```

Speed(114min per epoch):
```shell
Starting training for 300 epochs...
      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
      1/300      10.9G      2.424      3.916      2.939        198        640: 100%|██████████| 7402/7402 [1:53:59<00:00,  1.08it/s]
```

By using the cProfile tool to analyze performance, it was found that during the first epoch of approximately 7000 seconds, the clip's encode operation took over 2300 seconds:
<img width="1247" alt="Image" src="https://github.com/user-attachments/assets/8aa2d875-47f4-4e49-a2c8-7d6606e6eda6" />

# Improvement Example
The following demo is just a test example, it works, but it may not comply with the code specifications.
Modify the following code in file `ultralytics/models/yolo/world/train.py` :

```python
class WorldTrainer(yolo.detect.DetectionTrainer):
    def __init__(...):
        self._text_feats = {}

    def get_text_feats(self, texts: list[str], device, dtype) -> torch.Tensor:
        """Store the text that has appeared in memory to avoid duplicate calculations."""
        seen = set()
        new_texts = [
            text for text in texts
            if not (text in seen or seen.add(text))
               and text not in self._text_feats  
        ]

        if new_texts:
            new_tokens = self.clip.tokenize(new_texts).to(device)
            new_feats = self.text_model.encode_text(new_tokens).to(dtype=dtype)

            self._text_feats.update(zip(new_texts, new_feats))

        return torch.stack([self._text_feats[text] for text in texts], dim=0)

    def preprocess_batch(self, batch):
        """Preprocess a batch of images and text for YOLOWorld training."""
        batch = super().preprocess_batch(batch)

        # Add text features
        texts = list(itertools.chain(*batch["texts"]))
        #text_token = self.clip.tokenize(texts).to(batch["img"].device)
        #txt_feats = self.text_model.encode_text(text_token).to(dtype=batch["img"].dtype)  # torch.float32
        txt_feats = self.get_text_feats(texts, device=batch["img"].device, dtype=batch["img"].dtype)
        txt_feats = txt_feats / txt_feats.norm(p=2, dim=-1, keepdim=True)
        batch["txt_feats"] = txt_feats.reshape(len(batch["texts"]), -1, txt_feats.shape[-1])
        return batch
```

Speed(33min):
```Shell
Starting training for 300 epochs...
      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
      1/300      11.5G      2.399      4.222      2.914        168        640: 100%|██████████| 7402/7402 [32:23<00:00,  3.81it/s]
```

## Hit rate analysis
In the above dataset selection, over 99% of text features can be hit after 10 batches.

### Use case

Optimization can be achieved on most datasets without changing the code.

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @h13-0, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response to help guide you 🎯. An Ultralytics engineer will also review your feature suggestion and assist you soon!

**Comment by glenn-jocher** [comment]
Thank you for this excellent optimization suggestion! Caching text features is indeed a smart approach that could significantly improve YOLO World training performance. The 3.5x speedup you've demonstrated (114min → 33min per epoch) is impressive and would benefit many users working with large datasets containing repetitive text prompts.

Your analysis identifying that the CLIP encode operation consumes around 33% of the training time highlights a clear opportunity for optimization, especially since most datasets have a limited number of unique text prompts compared to their total samples.

The implementation you've suggested is clean and focused on the performance bottleneck. I'll bring this to the attention of our development team to consider incorporating into the core codebase. If you're interested in submitting a PR, that would be welcomed - it could include:

1. The caching mechanism you've demonstrated
2. Optional memory management for extremely large datasets (clearing cache periodically if needed)
3. Some basic documentation on the optimization

Thanks for your contribution to making YOLO World more efficient!

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in PR #20480, which introduces several improvements related to text feature caching for YOLOWorld and YOLOE training. This update optimizes how text embeddings are generated and stored, resulting in faster and more efficient training, especially for models that use text-based categories. Key enhancements include:

- Caching of text embeddings for category names to minimize redundant computations during training.
- Unified and improved logic for generating and storing text embeddings across different trainers.
- Modularized the process with a new build_text_model utility, replacing direct CLIP imports for better compatibility.
- More robust preprocessing and integration of text features into training batches.
- Improved caching mechanism using model-specific filenames, ensuring embeddings are only regenerated if necessary.
- Code cleanups and a version bump to 8.3.137.

Special thanks to @h13-0 for leading this PR, with valuable contributions from @glenn-jocher and @Laughing-q.

To test if this resolves the issue, you can upgrade to the latest version with:
```
pip install -U ultralytics>=8.3.137
```
Alternatively, you can install directly from the main branch with:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
The fix will also be available in the next official PyPI release.

Please let us know if these changes resolve your issue. Thank you for reporting this, and feel free to share any further feedback if the problem persists! 🙏
